### PR TITLE
Add syntax highlighting to the logged SQL

### PR DESCRIPTION
**Pull request by gavinking**
Using ANSI escape codes, we can make the logged SQL a little more readable.

For https://hibernate.atlassian.net/browse/HHH-14217

### CODE REVIEW

**Comment by gavinking** [comment]
Do we need some way to turn this off?

**Comment by NathanQingyangXu** [comment]
I really like the idea and this is the first PR I even cloned locally to inspect with curiosity. Currently it seems the formatting interferes with testing and most of the testing case failed (at the `drop table if exists ...` statement which has been changed to invalid SQL ending up with exception).
But this is a really cool feature!

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/internal/HighlightingFormatter.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * Performs basic syntax highlighting for SQL using ANSI escape codes.
+ *
+ * @author Gavin King
+ */
+public class HighlightingFormatter implements Formatter {
+
+    private static final Pattern keywords =
+            Pattern.compile("\\b(create|drop|alter|add|table|sequence|primary|foreign|unique|key|constraint|cascade|references|not|null|if|exists|select|as|delete|insert|update|set|from|in|into|values|left|right|inner|outer|join|on|where|having|order|group|by|desc|asc)\\b");
+    private static final Pattern strings =
+            Pattern.compile("'[^']*'");
```
Comment (NathanQingyangXu) [inline hibernate-core/src/main/java/org/hibernate/engine/jdbc/internal/HighlightingFormatter.java:15]: Guess we don't need to make everything perfect, but RE based matching might be broken for the following SQL:
```
update book b set b.title = 'Bob Dylan\'s dream' where b.id = 1;
```

Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/engine/jdbc/internal/HighlightingFormatter.java:15]: Yeah look, I'm not sure if it's really worth trying to make it completely perfect, but in principle I could ultimately get rid of the regexes and write a proper tokenizer for SQL. That's not hard.

Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/engine/jdbc/internal/HighlightingFormatter.java:15]: Oh and also the correct way to escape a ' in a SQL string is '', isn't it?

**Comment by gavinking** [comment]
So we have tests which look at the output of the logged SQL?

Because the formatters shouldn't be affecting the SQL sent to the database should they?

**Comment by gavinking** [comment]
> Because the formatters shouldn't be affecting the SQL sent to the database should they?

Oh for DDL the formatted SQL is executed?!

Is that ... *supposed* to work like that?

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/internal/HighlightingFormatter.java (hunk 1 above)
Comment (NathanQingyangXu) [reply hibernate-core/src/main/java/org/hibernate/engine/jdbc/internal/HighlightingFormatter.java]: Yeah. It is double single quotes.

**Comment by NathanQingyangXu** [comment]
> > Because the formatters shouldn't be affecting the SQL sent to the database should they?
> 
> Oh for DDL the formatted SQL is executed?!
> 
> Is that ... _supposed_ to work like that?

Saw you have fixed the testing case.

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/internal/HighlightingFormatter.java (hunk 1 above)
Comment (NathanQingyangXu) [reply hibernate-core/src/main/java/org/hibernate/engine/jdbc/internal/HighlightingFormatter.java]: Yeah, if it is really worth trying to make it perfect, we might well get formatting info from SQL generation phase, which might be more robust than RE. Anyway, if something imperfect is released, we might need to fix some bug reported by our users sooner or later.

Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/engine/jdbc/internal/HighlightingFormatter.java]: I've changed it to use a handwritten tokenizer.

**Comment by NathanQingyangXu** [comment]
It is definitively pretty cool and helpful for developer inspecting SQL in IDE, but I am not sure formatting is needed in all scenarios (like in high performance server side), so I think adding option to allow user to choose whether to format doesn't hurt. Maybe we can even go as far as disabling it by default, though personally I would enable this cool feature without any hesitation for sure.

**Review by NathanQingyangXu — APPROVED** [review]

**Comment by gavinking** [comment]
>  think adding option to allow user to choose whether to format doesn't hurt. 

Well I just checked and it seems that editors and even `less` and `more` don't know what to do with ANSI escape codes, and since there doesn't seem to be any way to automatically detect that we're logging to a file, I guess we will need a way to turn it off.

**Comment by gavinking** [comment]
> I guess we will need a way to turn it off

Alternatively I could limit it to be a feature of `hibernate.show_sql`.

**Comment by gavinking** [comment]
> > I guess we will need a way to turn it off
> 
> Alternatively I could limit it to be a feature of `hibernate.show_sql`.

I've added `hibernate.highlight_sql` to enable this. But I'm thinking that probably it should always be enabled for output in the console from `hibernate.show_sql`.

**Comment by gavinking** [comment]
Upon reflection, I've decided that this only really makes sense for the console output with `hibernate.show_sql`, and doesn't really make sense for the log. So for `hibernate.show_sql` I've turned it on by default, but let you explicitly disable it in case it's a problem.

**Comment by quaff** [comment]
It should ignore formatting if no console.
see `noConsoleNoAnsi=true` https://logging.apache.org/log4j/2.x/log4j-core/apidocs/org/apache/logging/log4j/core/pattern/HighlightConverter.html

**Comment by gavinking** [comment]
> It should ignore formatting if no console.

OK, good idea. I've made that change.

**Comment by gavinking** [comment]
> > It should ignore formatting if no console.
> 
> OK, good idea. I've made that change.

Actually, no, scratch that, that actually *doesn't* seem to be the right thing to do, since the console is actually `null` when running tests inside IntelliJ.

**Comment by quaff** [comment]
> @quaff
> 
> > > It should ignore formatting if no console.
> > 
> > 
> > OK, good idea. I've made that change.
> 
> Actually, no, scratch that, that actually _doesn't_ seem to be the right thing to do, since the console is actually `null` when running tests inside IntelliJ.

It will be mess in eclipse <strong>or</strong> windows cmd, maybe there is no better way.

**Comment by gavinking** [comment]
As I understand it, windows terminal now supports ANSI escapes.

I didn't know about the Eclipse console, does it really not?!

**Comment by gavinking** [comment]
> I didn't know about the Eclipse console, does it really not?!

Ugh. Apparently you need a plug-in :(

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/spi/SqlStatementLogger.java [hunk 2]
```diff
@@ -120,14 +133,16 @@ public void logStatement(String statement) {
...
+				&& ( logToStdout || LOG.isDebugEnabled() ) ) {
+			statement = formatter.format( statement );
 		}
 		LOG.debug( statement );
 		if ( logToStdout ) {
-			System.out.println( "Hibernate: " + statement );
+			if ( highlight ) {
+				statement = FormatStyle.HIGHLIGHT.getFormatter().format( statement );
+			}
+			System.out.println( "\u001b[35m[Hibernate]\u001b[0m " + statement );
```
Comment (quaff) [inline hibernate-core/src/main/java/org/hibernate/engine/jdbc/spi/SqlStatementLogger.java:145]: If not highlight you should not print Ansi escapes.

**Review by quaff — COMMENTED** [review]
> As I understand it, windows terminal now supports ANSI escapes.
> 
> I didn't know about the Eclipse console, does it really not?!

Windows `cmd` doesn't supports ANSI escapes, I really like this feature, but I think it should be disabled by default.

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/spi/SqlStatementLogger.java (hunk 2 above)
Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/engine/jdbc/spi/SqlStatementLogger.java]: Fixed.

**Comment by gavinking** [comment]
> Windows `cmd` doesn't supports ANSI escapes

Fine, so it's now disabled by default.

**Comment by Sanne** [comment]
Very nice!  About the default, I believe maybe Quarkus already has a general way to know if the output supports it?  I hope we could tie into that so to switch it on by default in Quarkus when it hints this won't break things?

@maxandersen , @gsmet ?

**Comment by maxandersen** [comment]
There isn't really a reliable way to know. 

Currently we disable quarkus ansi logging if running on windows no matter if there is a console or not. This is done very early on before any config set/configured if I recall. 

I agree we should try and see if those flags can be matched / aligned.

**Comment by Sanne** [comment]
thanks @maxandersen . I suppose that if Quarkus is emitting ansi codes, it's fine if Hibernate does so as well; to be continued on the Quarkus tracker ;)

File: documentation/src/main/asciidoc/userguide/appendices/Configurations.adoc [hunk 3]
```diff
@@ -621,6 +621,9 @@ Write all SQL statements to the console. This is an alternative to setting the l
 `*hibernate.format_sql*` (e.g. `true` or `false` (default value))::
 Pretty-print the SQL in the log and console.
 
+`*hibernate.highlight_sql*` (e.g. `true` (default value) or `false`)::
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/userguide/appendices/Configurations.adoc:624]:
```suggestion
`*hibernate.highlight_sql*` (e.g. `true` or `false` (default value))::
```

Comment (Sanne) [reply documentation/src/main/asciidoc/userguide/appendices/Configurations.adoc]: thanks @yrodiere !

**Comment by Sanne** [comment]
Merged and backported. Thanks!

**Comment by gavinking** [comment]
@Sanne thanks.

**Pull request closed without merging**